### PR TITLE
Writes nodegraph uifolder as well as copies uifolder to nodedef and

### PR DESCRIPTION
**Pull request by feldstj**
writes it out.

### CODE REVIEW

**Comment by feldstj** [comment]
Currently uifolder is being written out if added to a nodegraph or nodedef. Also the nodegraph input uifolders are copied to the nodedef inputs for publising. If you prefer this be generalized to all metadata being written out, let me know. I was just using this for my specific needs.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 1]
```diff
@@ -715,6 +715,15 @@ namespace
                 else
                 {
                     destPort = destNodeDef->addInput(attr->getName(), attr->getType().str());
+                    const RtTypedValue* uiFolderMetaData = input->getMetadata(RtToken(ValueElement::UI_FOLDER_ATTRIBUTE));
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp:718]: Write out the nodedef input uifolders

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 2]
```diff
@@ -937,6 +946,15 @@ namespace
             else
             {
                 InputPtr input = destNodeGraph->addInput(nodegraphInput.getName(), nodegraphInput.getType());
+                RtTypedValue* uiFolderMetaData = nodegraphInput.getMetadata(RtToken(ValueElement::UI_FOLDER_ATTRIBUTE));
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp:949]: Write out the nodegraph input uifolders.

File: source/MaterialXRuntime/RtStage.cpp [hunk 3]
```diff
@@ -171,10 +173,16 @@ RtPrim RtStage::createNodeDef(RtNodeGraph& nodeGraph,
         nodedef.setNodeGroup(nodeGroup);
     }
 
-    // Add an output per nodegraph input
+    // Add an input per nodegraph input
     for (auto input : nodeGraph.getInputs())
     {
-        RtAttribute attr = nodedef.createInput(input.getName(), input.getType());
+        RtInput attr = nodedef.createInput(input.getName(), input.getType());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtStage.cpp:179]: Copy the nodegraph input uifolders to the nodedef inputs if there are any. This is needed for publishing.

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 1 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp:718]: Should we not copy over all ui information in this case. Not sure why just ui folder?

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 2 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtFileIo.cpp:949]: Seems like we should just write out all meta data.

File: source/MaterialXRuntime/RtStage.cpp (hunk 3 above)
Comment (bernardkwok) [reply source/MaterialXRuntime/RtStage.cpp:179]: Again why just ui folder?

File: source/MaterialXRuntime/RtFileIo.cpp (hunk 1 above)
Comment (niklasharrysson) [reply source/MaterialXRuntime/RtFileIo.cpp:718]: Hmm.. exporting all metadata should already be handled by the call to writeMetadata above (line 701). Is that not working?

File: source/MaterialXRuntime/RtStage.cpp [hunk 4]
```diff
@@ -11,6 +11,8 @@
 
 #include <MaterialXRuntime/Private/PvtStage.h>
 
+#include <MaterialXCore/Element.h>
```
Comment (niklasharrysson) [inline source/MaterialXRuntime/RtStage.cpp:14]: Is this include needed?

File: source/MaterialXRuntime/RtStage.cpp (hunk 3 above)
Comment (niklasharrysson) [reply source/MaterialXRuntime/RtStage.cpp:179]: Yes, you can copy all metadata instead. Accessible from the private API:

```
const PvtObject* obj = PvtObject::hnd(attr)->asA<PvtObject>();
const vector<RtToken>& metadataNames = obj->getMetadataOrder();
for (auto metadataName : metadataNames)
{
    const RtTypedValue* metadataValue = obj->getMetadata(metadataName);
    ... 
}
```

File: source/MaterialXRuntime/RtStage.cpp (hunk 4 above)
Comment (feldstj) [reply source/MaterialXRuntime/RtStage.cpp:14]: It was needed for the UI folder constant, but now that that has been removed, I'll remove this.

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 5]
```diff
@@ -949,6 +980,10 @@ namespace
                     }
                 }
             }
+            if (v)
+            {
+                v->setValueString(nodegraphInput.getValueString());
```
Comment (feldstj) [inline source/MaterialXRuntime/RtFileIo.cpp:985]: Writes out input values

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 6]
```diff
@@ -715,6 +715,20 @@ namespace
...
+                    const vector<RtToken>& metadataNames = input->getMetadataOrder();
+                    for (auto metadataName : metadataNames)
+                    {
+                        const RtTypedValue* metadataValue = input->getMetadata(metadataName);
+                        if (metadataValue)
+                        {
+                            std::string metadataString = metadataValue->getValueString();
+                            if (!metadataString.empty())
+                            {
+                                destPort->setAttribute(metadataName, metadataString);
```
Comment (niklasharrysson) [inline source/MaterialXRuntime/RtFileIo.cpp:728]: About this loop to write metadata, I think the call to `writeMetadata(attr, destPort, attrMetadata);` below (line 754) should already handle writing this. Is it not?

File: source/MaterialXRuntime/RtFileIo.cpp [hunk 7]
```diff
@@ -930,13 +944,30 @@ namespace
...
+                const vector<RtToken>& metadataNames = obj->getMetadataOrder();
+                for (auto metadataName : metadataNames)
+                {
+                    const RtTypedValue* metadataValue = obj->getMetadata(metadataName);
+                    if (metadataValue)
+                    {
+                        std::string metadataString = metadataValue->getValueString();
+                        if (!metadataString.empty())
+                        {
+                            input->setAttribute(metadataName, metadataString);
```
Comment (niklasharrysson) [inline source/MaterialXRuntime/RtFileIo.cpp:967]: I think this could be done by just calling `writeMetadata` here as well.

**Pull request merged**